opentelemetry-exporter-geneva/geneva-uploader-ffi/src/lib.rs

```diff
@@ -1 +1,1027 @@
+///
+/// # Safety
+/// - handle must be a valid pointer returned by geneva_client_new
+/// - data must be a valid pointer to protobuf-encoded ExportLogsServiceRequest
+/// - data_len must be the correct length of the data
+/// - out_batches must be non-null; on success it receives a non-null pointer the caller must free with geneva_batches_free
+#[no_mangle]
+pub unsafe extern "C" fn geneva_encode_and_compress_logs(
+    handle: *mut GenevaClientHandle,
+    data: *const u8,
+    data_len: usize,
+    out_batches: *mut *mut EncodedBatchesHandle,
+) -> GenevaError {
+    if out_batches.is_null() {
+        return GenevaError::NullPointer;
+    }
+    unsafe { *out_batches = ptr::null_mut() };
+
+    if handle.is_null() {
+        return GenevaError::NullPointer;
+    }
+    if data.is_null() {
+        return GenevaError::NullPointer;
+    }
+    if data_len == 0 {
+        return GenevaError::EmptyInput;
+    }
+
+    // Validate handle first
+    let validation_result = unsafe { validate_handle(handle) };
+    if validation_result != GenevaError::Success {
+        return validation_result;
+    }
+
+    let handle_ref = match unsafe { handle.as_ref() } {
+        Some(h) => h,
+        None => return GenevaError::NullPointer,
+    };
+    let data_slice = unsafe { std::slice::from_raw_parts(data, data_len) };
+
+    let logs_data: ExportLogsServiceRequest = match Message::decode(data_slice) {
+        Ok(data) => data,
+        Err(_e) => {
+            return GenevaError::DecodeFailed;
+        }
+    };
+
+    let resource_logs = logs_data.resource_logs;
+    match handle_ref.client.encode_and_compress_logs(&resource_logs) {
+        Ok(batches) => {
+            let h = EncodedBatchesHandle {
+                magic: GENEVA_HANDLE_MAGIC,
+                batches,
+            };
+            unsafe { *out_batches = Box::into_raw(Box::new(h)) };
+            GenevaError::Success
+        }
+        Err(_e) => GenevaError::InternalError,
+    }
+}
+
+/// Returns the number of batches in the encoded batches handle
+///
+/// # Safety
+/// - batches must be a valid pointer returned by geneva_encode_and_compress_logs, or null
+#[no_mangle]
+pub unsafe extern "C" fn geneva_batches_len(batches: *const EncodedBatchesHandle) -> usize {
+    // Validate batches
+    match unsafe { validate_handle(batches) } {
+        GenevaError::Success => {
+            // Safe to dereference after validation
+            let batches_ref = unsafe { batches.as_ref().unwrap() };
+            batches_ref.batches.len()
+        }
+        _ => 0, // Return 0 for invalid handles
+    }
+}
+
+/// Uploads a specific batch synchronously
+///
+/// # Safety
+/// - handle must be a valid pointer returned by geneva_client_new
+/// - batches must be a valid pointer returned by geneva_encode_and_compress_logs
+/// - index must be less than the value returned by geneva_batches_len
+#[no_mangle]
+pub unsafe extern "C" fn geneva_upload_batch_sync(
```

Comment:
Also, could we offer an iterator based API instead of getting tied to an index based API?